File: Data Structures & Algorithms/best-time-to-buy-and-sell-stock-ii/submission-3.py
class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        n = len(prices)
        res = 0
        help = float("inf")
        for num in prices:
            if num < help:
                help = num
            else:
                res += num - help
                help = num
        return res

File: Data Structures & Algorithms/buy-and-sell-crypto-with-cooldown/submission-3.cpp
class Solution {
public:
    int maxProfit(vector<int>& prices) {
        return dp(prices, 0, 0);
    }
    int dp(vector<int>& prices, int i, int buy) {
        if (i >= prices.size())
            return 0;
        if (buy == 0) {
            return std::max(-prices[i] + dp(prices, i + 1, 1), dp(prices, i + 1, 0));
        }
        else {
            return std::max(+prices[i] + dp(prices, i + 2, 0), dp(prices, i + 1, 1));
        }
    }
};
